keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/HamiltonianSampler.java

```diff
@@ -0,0 +1,280 @@
+    }
+
+    @Override
+    public void step() {
+        initializeMomentumForEachVertex(latentVertices, during.momentum, random);
+        cacheThings(during, before);
+        putSample(sampleBeforeLeapfrog, fromVertices);
+
+        for (int leapFrogNum = 0; leapFrogNum < leapFrogCount; leapFrogNum++) {
+            during.gradient = leapfrog(
+                latentVertices,
+                during,
+                stepSize,
+                logProbGradientCalculator
+            );
+        }
+    }
+
+    @Override
+    public void sample(Map<VertexId, List<?>> samples, List<Double> logOfMasterPForEachSample) {
+        step();
+        addSampleFromCache(samples, acceptOrReject());
+        logOfMasterPForEachSample.add(logOfMasterPBeforeLeapfrog);
+    }
+
+    @Override
+    public NetworkState sample() {
+        step();
+        return new SimpleNetworkState(acceptOrReject());
+    }
+
+    private Map<VertexId, ?> acceptOrReject() {
+        final double logOfMasterPAfterLeapfrog = bayesNet.getLogOfMasterP();
+
+        final double likelihoodOfLeapfrog = getLikelihoodOfLeapfrog(
+            logOfMasterPAfterLeapfrog,
+            logOfMasterPBeforeLeapfrog,
+            during.momentum,
+            before.momentum
+        );
+
+        if (shouldReject(likelihoodOfLeapfrog, random)) {
+
+            //Revert to position and gradient before leapfrog
+            Map<VertexId, DoubleTensor> tempSwap = during.position;
+            during.position = before.position;
+            before.position = tempSwap;
+
+            tempSwap = during.gradient;
+            during.gradient = before.gradient;
+            before.gradient = tempSwap;
+
+            return sampleBeforeLeapfrog;
+        } else {
+            logOfMasterPBeforeLeapfrog = logOfMasterPAfterLeapfrog;
+            Map<VertexId, ?> sample = new HashMap<>();
+            putSample(sample, fromVertices);
+            return sample;
+        }
+    }
+
+    private static void initializeMomentumForEachVertex(List<Vertex<DoubleTensor>> vertexes,
+                                                        Map<VertexId, DoubleTensor> momentums,
+                                                        KeanuRandom random) {
+        for (Vertex<DoubleTensor> currentVertex : vertexes) {
+            momentums.put(currentVertex.getId(), random.nextGaussian(currentVertex.getShape()));
+        }
+    }
+
+    private static void cacheThings(VertexState from, VertexState to) {
+        for (Map.Entry<VertexId, DoubleTensor> entry : from.position.entrySet()) {
+            to.position.put(entry.getKey(), entry.getValue());
+        }
+        for (Map.Entry<VertexId, DoubleTensor> entry : from.momentum.entrySet()) {
+            to.momentum.put(entry.getKey(), entry.getValue());
+        }
+        for (Map.Entry<VertexId, DoubleTensor> entry : from.gradient.entrySet()) {
+            to.gradient.put(entry.getKey(), entry.getValue());
+        }
```

Comment:
to.position.putAll(from.position);
        to.momentum.putAll(from.momentum);
        to.gradient.putAll(from.gradient);

Reply:
This feels like something that should be encapsulated as a method in `VertexState`

Reply:
Nice!